bindings/nodejs/src/functions/uart.cc

```diff
@@ -32,19 +34,63 @@ class SolUART : public JSHandle<SolUART> {
     static const char *jsClassName() { return "SolUART"; }
 };
 
-static void sol_uart_read_callback(void *user_data, struct sol_uart *uart,
-    unsigned char byte_read) {
+static ssize_t sol_uart_on_data_callback(void *user_data,
+    struct sol_uart *uart, const struct sol_buffer *buf) {
     Nan::HandleScope scope;
     sol_uart_data *uart_data = (sol_uart_data *)user_data;
-    Nan::Callback *callback = uart_data->rx_cb;
+    if (!uart_data)
+        return 0;
+
+    Nan::Callback *callback = uart_data->on_data_cb;
     if (!callback)
-        return;
+        return 0;
+
+    struct sol_str_slice slice;
+    slice = sol_buffer_get_slice(buf);
+    Local<Value> buffer =
+        Nan::CopyBuffer(slice.data, slice.len).ToLocalChecked();
```

Comment:
the idea of passing a buffer is to improve memory usage and copies. I know the user may take a reference to that buffer and thus you'd need to copy. But can you do that on demand? Like "CopyOnReference", that would only copy the given slice if there is a reference taken when the function returns.

As this is a wrapper object, you can easily replace its internals, so a way to do that would be to check after `callback->Call(1, arguments)`, if `buffer` reference is 1 and would die. In this case you'd introduce something like `buffer->ForceCopy()`

We'll use this pattern in other places, like the HTTP progressive response (likely not be wrapped in nodejs, since you already have that as a native module), Crypto/message-digest and so on.